File: web-backend/src/main/java/com/github/kr328/webapi/tools/BurstLimiter.java
package com.github.kr328.webapi.tools;

import lombok.AllArgsConstructor;

import java.util.HashMap;

public class BurstLimiter {
    private HashMap<String, Record> table;

    public BurstLimiter() {
        table = new HashMap<>();
    }

    public synchronized boolean note(String id) {
        Record record = table.computeIfAbsent(id, s -> new Record(0, System.currentTimeMillis()));

        if (System.currentTimeMillis() - record.last > 1000 * 60)
            record.count = 0;

        record.last = System.currentTimeMillis();

        return record.count++ > 20;
    }

    @AllArgsConstructor
    private static class Record {
        private int count;
        private long last;
    }
}

File: web-backend/src/main/java/com/github/kr328/webapi/tools/BurstLimiter.kt
package com.github.kr328.webapi.tools

import kotlin.concurrent.thread
import kotlin.concurrent.timer

class BurstLimiter {
    private val table: MutableMap<String, Record> = mutableMapOf()

    init {
        timer(period = 60 * 1000) { cleanup() }
    }

    @Synchronized
    fun note(id: String): Boolean {
        val record = table.computeIfAbsent(id) { Record(0, System.currentTimeMillis()) }

        if (System.currentTimeMillis() - record.last > 1000 * 60)
            record.count = 0

        record.last = System.currentTimeMillis()

        return record.count++ > 20
    }

    @Synchronized
    fun cleanup() {
        table.entries.removeIf { System.currentTimeMillis() - it.value.last > 60 * 1000 }
    }

    private data class Record(var count: Long, var last: Long)
}
